errors/multi.go

```diff
@@ -1,125 +1,118 @@
 package errors
 
 import (
-	"encoding/json"
 	"fmt"
 	"strings"
+	"unicode"
+
+	"github.com/pkg/errors"
 )
 
-// Append combines two errors. Any value can be nil.
-// When both are nil then nil is returned, too.
-// When only one is not nil, that one is returned.
-// Otherwise the result is of type multiErr.
-func Append(src, new error) error {
-	switch {
-	case new == nil:
-		return src
-	case src == nil:
-		return new
+// Append clubs together all provided errors. Nil values are ignored.
+func Append(errs ...error) error {
+	// Always build the multi error collection from scratch to avoid slice
+	// modyfications.
+	var res multiError
+
+	for _, e := range errs {
+		res = appendError(res, e)
 	}
-	return multiErr{}.with(src).with(new)
+	if len(res) == 0 {
+		return nil
+	}
+	return res
 }
 
-const multiErrCode uint32 = 1000
-
-var _ coder = (*multiErr)(nil)
-
-type multiErr []error
-
-// IsEmpty returns true if there are no errors registered,
-func (e multiErr) IsEmpty() bool {
-	return len(e) == 0
-}
+// appendError extends given multiError with provided error. It flattens any
+// error that provides the Unpack method.
+func appendError(errs multiError, e error) multiError {
+	if isNilErr(e) {
+		return errs
+	}
 
-// with returns a new multiErr instance with the given error added.
-// Nil values are ignored and multiErr flattened.
-func (e multiErr) with(source error) multiErr {
-	switch err := source.(type) {
-	case nil:
-		return e
-	case multiErr:
-		return e.append(err...)
-	case *wrappedError:
-		root, msgs := unWrap(err)
-
-		me, ok := root.(multiErr)
-		if !ok {
-			return e.append(source)
-		}
-		// flatten and re-wrap errors
-		result := e
-		for _, v := range me {
-			rErr := v
-			for _, m := range msgs {
-				rErr = Wrap(rErr, m)
-			}
-			result = append(result, rErr)
+	if u, ok := e.(unpacker); ok {
```

Comment:
This is a very good point. It requires a decision on how to handle such a case, as you cannot flatten this error.

I have decided to not flatten it and preserve the wrap, as I believe this is the only non desctructive solution.

I have **not** decided to flatten the error as this would destroy the wrapping. Expectation might be that 
```
Append(Wrap(Append(foo, bar), "context")
``` 
is equal to 
```
Append(Wrap(foo, "context"), Wrap(bar, "context))
```
 but this is a wrong assumtion. Those two constructs are not equal as you do not know what `Wrap` does. For example, you cannot reassign the stack trace it carries. This problem would occur with any other error wrapper/constructor, for example `Field`.

This is one of the changes and differences with the previous implementation. Errors relation is a tree not a flat list. There is only a very limited set of cases when a multi-error can be flattened. For complex constructs there is no other way as to agree to the tree structure.

Does this make sense?

Reply:
Yeah, this does make total sense to me too. I like how this works without flattening.

Reply:
I agree also with the approach. Just to make it a bit clearer in the docs.
It is a very important design decision. You made a good decision. Just make sure people understand why in 6 months